zaza/openstack/charm_tests/openstack_dashboard/tests.py

```diff
@@ -206,6 +208,64 @@ def test_100_services(self):
                 target_status='running'
             )
 
+    def test_200_haproxy_stats_config(self):
+        """Verify that the HAProxy stats are properly setup."""
+        logging.info('Checking dashboard HAProxy settings...')
+        application_name = 'openstack-dashboard'
+        unit = zaza_model.get_unit_from_name(
+            zaza_model.get_lead_unit_name(application_name))
+        keystone_unit = zaza_model.get_lead_unit_name('keystone')
+        dashboard_relation = openstack_juju.get_relation_from_unit(
+            keystone_unit, unit.entity_id, 'identity-service')
+        dashboard_ip = dashboard_relation['private-address']
+        logging.debug("... dashboard_ip is:{}".format(dashboard_ip))
+        conf = '/etc/haproxy/haproxy.cfg'
+        port = '8888'
+        set_alternate = {
+            'haproxy-expose-stats': 'True',
+        }
+
+        request = urllib.request.Request(
+            'http://{}:{}'.format(dashboard_ip, port))
+
+        # NOTE(ajkavanagh): it seems that apache2 doesn't start quickly enough
+        # for the test, and so it gets reset errors; repeat until either that
+        # stops or there is a failure
+        @tenacity.retry(wait=tenacity.wait_exponential(multiplier=1,
+                                                       min=5, max=10),
+                        retry=tenacity.retry_if_exception_type(
+                            http.client.RemoteDisconnected),
+                        reraise=True)
+        def _do_request():
+            return urllib.request.urlopen(request)
```

Comment:
This all seems to be duplicate code. So we should generalize a helper and use the tenacity decorator on the helper itself.